[FIX JENKINS-43148] Pipeline editor load/edit/save with feature flag

kzantow opened the pull request:
# Description

Git editing with feature flag; there will be follow-on PRs to enable it entirely, this is to begin to enable testing in beta. To enable this, add the system property `blueocean.features.GIT_PIPELINE_EDITOR=true` e.g. from the command line, it's: `-Dblueocean.features.GIT_PIPELINE_EDITOR=true`. This is currently functional if the git project is created with a sufficiently privileged credential

Here's a video of this in action:

[![Git Creation Flow](https://img.youtube.com/vi/TUIz7t_CMnM/0.jpg)](https://www.youtube.com/watch?v=TUIz7t_CMnM)

See [JENKINS-43148](https://issues.jenkins-ci.org/browse/JENKINS-43148).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

jglick commented on blueocean-git-pipeline/pom.xml:
```diff
@@ -24,8 +24,13 @@
+    <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>git</artifactId>
+      <version>3.5.1-SNAPSHOT</version>
```
Better to `mvn -DskipTests clean deploy` a snapshot and use the timestamped version here. Otherwise your CI build is at the mercy of anyone else who might have been messing around with unrelated patches to the upstream plugin.

kzantow replied:
@jglick right, I'm just waiting for a new git-plugin release, should be this weekend, I'm told - the people who need to can't easily test this until the git-plugin is published at least to the experimental update center, anyway AFAIK

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java:
```diff
@@ -0,0 +1,286 @@
+        GitReadSaveRequest r = null;
+        try {
+            r = makeSaveRequest(item, req);
+            r.cloneRepo();
```
So every time we load editor we are cloning the repo? Git pipeline might point to a remote Github/GHE repo as well, so that might be expensive cloning every time? Maybe worth checking with @stephenc, he implemented shallow cloning with github/bitbucket and maybe Git.

kzantow replied:
We're cloning a local filesystem repo, I'll make sure it's shallow cloned. Also experimenting with just operating on a bare repository directly, found some ways to do that (still with a bit of drawbacks)

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitPipelineUpdateRequest.java:
```diff
@@ -0,0 +1,137 @@
+/**
+ * @author Vivek Pandey
+ */
+public class GitPipelineUpdateRequest extends BluePipelineCreateRequest {
```
No, we just need the create request, it's a byproduct of the merge I didn't want to remove until I can have it all working without it.

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitContent.java:
```diff
@@ -0,0 +1,192 @@
+/**
+ * @author Vivek Pandey
+ */
+public class GitContent extends ScmContent {
```
You're probably right, I'll make sure there's not duplicated stuff unless it's necessary

kzantow commented:
@vivek thanks for the feedback, yes, this is still a WiP... I want to get it merged this way (after a few changes addressing your comments) so people can test it (we will find a lot of issues this way, I'm sure since people's Git setups are all over the place), it's all hidden behind a feature flag.

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java:
```diff
@@ -0,0 +1,286 @@
+        byte[] read() throws IOException, InterruptedException {
+            try {
+                // thank you test for how to use something...
+                // https://github.com/jenkinsci/git-client-plugin/blob/master/src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java#L1108
```
Remove comments to example code that someone might need to use in the future? I'd rather not.

michaelneale commented:
Not sure if indicative of a problem but I did see this in console when opening editor for a ssh git pipeline created with system cred: 

```
TypeError: Cannot read property 'self' of undefined
    at ActivityService.bunkerKey (blueocean.js:5598)
    at ActivityService.setItem (blueocean.js:5641)
    at executeAction (blueocean.js:65191)
    at ActivityService.res (blueocean.js:65163)
    at blueocean.js:5892
    at <anonymous>
```

michaelneale commented:
so I tried it with "gogs" - can save if I use system cred only. 

I tried with http to checkout, and added the generated public key - but that caused an error 

SEVERE: Error running git remote-ls: org.eclipse.jgit.api.errors.TransportException: git@localhost:michaelneale/testy.git: Failed to connect
hudson.plugins.git.GitException: org.eclipse.jgit.api.errors.TransportException: git@localhost:michaelneale/testy.git: Failed to connect

(maybe I did that wrong). 

I did see one :bug: which is with an empty repo - it doesn't offer to create a pipeline the first time like github does (you may already know this).

i386 commented:
That makes sense though. Can't use the public key with HTTPS. The design removes the ability to use HTTPS - we should validate that input field.

michaelneale commented:
@kzantow is the findbugs build failure legit?

michaelneale commented:
findbugs still grumpy it seems - this uses published versions of things or beta versions of it and editor?

i386 commented on blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/UserImpl.java:
```diff
@@ -155,41 +155,41 @@ public Link getLink() {
             UserSSHKeyManager.getOrCreate(authenticatedUser));
-        
+
         return HttpResponses.okJSON(ImmutableMap.of("key", publicKey));
```
Could this return its own model like other methods in our API?
``
@ExportedBean
public class UserKey {
  private final publicKey;

  @Exported(name="publickey")
  public String getPublicKey() {
    return publicKey;
  }
}

```

kzantow replied:
Yes, done

i386 commented on blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/UserImpl.java:
```diff
@@ -155,41 +155,41 @@ public Link getLink() {
+
         UserSSHKeyManager.reset(authenticatedUser);
         return HttpResponses.ok();
     }
```
Should return `UserKey` model of the new key.

michaelneale commented:
OK I tried this: 

0) setup a local repo with "gogs"
1) setup mic user and setup MBP - builds ok
2) created user "alice"
3) took alice public key and made gogs trust it
4) Got this error when alice tries to save: 

```
Caused by: org.eclipse.jgit.errors.TransportException: git@localhost:michaelneale/testy.git: UnknownHostKey: localhost. RSA key fingerprint is db:d1:03:95:3f:2a:b6:8f:1d:af:6f:b8:98:33:07:ed
	at org.eclipse.jgit.transport.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:160)
	at org.eclipse.jgit.transport.SshTransport.getSession(SshTransport.java:137)
	at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:264)
	at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:162)
	at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136)
	at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122)
	at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1201)
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:128)
	... 109 more
```

kzantow commented:
@i386 your comments addressed. @michaelneale I think this should be functional for your case where you haven't connected via ssh from the command line; tests are passing without having a host added to the local `known_hosts`

kzantow commented:
@vivek @i386 @michaelneale hm, I'm trying to test connecting to a repository over ssh, but here I see: **ssh: not found**, any ideas how I should go about testing this? (it _does_ work on my machine)

michaelneale commented:
I am testing with https://github.com/gogits/gogs/tree/master/docker - if that helps. Maybe github cold be fine... might have to walk us through this tomorrow...

michaelneale commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitUtils.java:
```diff
@@ -128,7 +128,7 @@ private static TransportConfigCallback getSSHKeyTransport(final BasicSSHUserPriv
             @Override
             protected void configure(OpenSshConfig.Host hc, com.jcraft.jsch.Session session) {
-                // do nothing
+                session.setConfig("StrictHostKeyChecking", "no"); // jenkins user doesn't likely have the hos key
```
Nor would Jenkins have the host key

kzantow commented:
Hm, I just tested this out with Gogs and it works. One issue is I don't think the right keys are showing up in the initial selector. I had to create the repo with default, but then saving appropriately used the user's personal key

michaelneale commented:
OK, can't quite work this out - user error somewhere. Can talk this over tomorrow to work out what expected behavior should be.

michaelneale commented:
@kzantow java.util.function.Supplier is a java8 only thing - I didn't think we had java8 stuff like that allowed? My intelliJ is barfing, I can't run the new test case either... so something is up

kzantow commented:
Hm. Really strange. It was an IDEA suggestion :)

michaelneale commented:
noted that no ssh command was actually in the docker image - let's see if this latest change effects things - maybe this is what @imeredith saw too

michaelneale commented:
ok @kzantow ssh thing seems fixed anyway. Turns out the error message was perfectly accurate.

michaelneale commented:
Nice one @kzantow I tested this with the "bob" and "alice" scenarios with the correct git ssh url for gogs, and works great. 

Given this is feature flagged - up to you to get it in. To make it complete - a UI that prompts the user when it fails to load with the public key would be good (instead of generic error message). 

However, before merging, I would like a bit of a close look by @vivek and @cliffmeyers not for functionality, but from a security perspective - if one of them can sign off this can go in any time with feature flag, or with the UI change to show the user the public key. 

Nice! this seems functional to me in the permutations I tried.

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        try {
+            GitReadSaveRequest r = makeSaveRequest(item, req);
```
@kzantow how does it protect alice not to use bob's content? Looking at code inside `makeSaveRequest()` and `GitReadSaveRequest`, looks like its being picked up from the multi-branch pipeline. We should be picking up the credential of the logged in user or User.current() and then setup git tool with it's ssh keys. Same question stands for `saveContent()`.

kzantow replied:
@vivek yes, those aren't fully implemented, I want to have `GitBareRepoReadSaveRequest` be the only option, but it needs to be tested with different people's setups, since it does some wacky stuff with the bare repo; I'll get to them in the next couple days

i386 commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitSCMReadSaveService.java:
```diff
@@ -81,7 +80,6 @@ void save() throws IOException, InterruptedException, GitException, URISyntaxExc
                 public Void invoke(Repository repository) throws IOException, InterruptedException {
                     Git activeRepo = getActiveRepository(repository);
                     File repoDir = activeRepo.getRepository().getDirectory().getParentFile();
-                    System.out.println("Repo cloned to: " + repoDir.getCanonicalPath());
```
maybe `log.debug()` ?

i386 commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCacheCloneReadSaveRequest.java:
```diff
@@ -0,0 +1,224 @@
+                    Git activeRepo = getActiveRepository(repository);
+                    File repoDir = activeRepo.getRepository().getDirectory().getParentFile();
+                    try {
+//                        if (!sourceBranch.equals(branch)) {
```
Needed? If not delete

i386 commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitUtils.java:
```diff
@@ -13,6 +17,9 @@
 import io.jenkins.blueocean.rest.impl.pipeline.credential.BlueOceanDomainRequirement;
+import java.io.ByteArrayInputStream;
+
+import org.eclipse.jgit.transport.*;
```
🐛 `*` import

i386 commented:
Nice tests 👍

kzantow commented on Dockerfile.build:
```diff
@@ -77,3 +77,9 @@ WORKDIR /home/bouser
+#========================================
+# Need ssh
+#========================================
+RUN sudo apt-get update -qqy && sudo apt-get install -y ssh
```
Added: https://issues.jenkins-ci.org/browse/JENKINS-46478

michaelneale commented:
ok have tried it with github - the multibranch github org. Some :bug:

1) should it always ask the user to add the public key? what if it already works with the system one? or they don't want write access? or is open source? (James may have already answered this)
2) I tried to add a repo with no Jenkinsfile - it allowed me to open editor, but editor failed to load: 

```
There was an error loading the pipeline from the Jenkinsfile in this repository. Correct the error by editing the Jenkinsfile using the declarative syntax then commit it back to the repository.
 
Cannot read property 'indexOf' of undefined
``` 
![screen shot 2017-09-05 at 12 18 01 pm](https://user-images.githubusercontent.com/14976/30042395-4ecd41ca-9234-11e7-8f85-8dfd16f6a64e.png)


(stack trace at bottom)

stack trace on server. the repo was: https://github.com/multibranchorg/justreadme - and I added my key

3) If I revoke my key later - will editor prompt me to add the key to github? 
4) when I added git@github.com:multibranchorg/app-store-demo.git - it didn't seem ti index for a long long time (I guess git clone and index for a larger repo does that - activity screen looked blank for a while)


The stack trace: 

```
[WARNING] Error while serving http://localhost:8080/jenkins/blue/rest/organizations/jenkins/pipelines/app-store-demo/scm/content
java.lang.reflect.InvocationTargetException
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:347)
	at org.kohsuke.stapler.ForwardingFunction.invoke(ForwardingFunction.java:63)
	at io.jenkins.blueocean.commons.stapler.TreeResponse$Processor.invoke(TreeResponse.java:43)
	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:184)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:117)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:129)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$3.doDispatch(MetaClass.java:209)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$3.doDispatch(MetaClass.java:209)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:686)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:845)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:225)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationFilter.doFilter(JwtAuthenticationFilter.java:51)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at jenkins.metrics.impl.MetricsFilter.doFilter(MetricsFilter.java:125)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:86)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Cannot open session, connection is not authenticated.
	at com.trilead.ssh2.Connection.openSession(Connection.java:1127)
	at org.jenkinsci.plugins.gitclient.trilead.TrileadSession.exec(TrileadSession.java:32)
	at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:264)
	at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:162)
	at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136)
	at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122)
	at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1201)
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:128)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitUtils.fetch(GitUtils.java:213)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitBareRepoReadSaveRequest$1.invoke(GitBareRepoReadSaveRequest.java:70)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitBareRepoReadSaveRequest$1.invoke(GitBareRepoReadSaveRequest.java:59)
	at jenkins.plugins.git.GitSCMFileSystem$3.invoke(GitSCMFileSystem.java:193)
	at org.jenkinsci.plugins.gitclient.AbstractGitAPIImpl.withRepository(AbstractGitAPIImpl.java:29)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.withRepository(CliGitAPIImpl.java:71)
	at jenkins.plugins.git.GitSCMFileSystem.invoke(GitSCMFileSystem.java:189)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitBareRepoReadSaveRequest.read(GitBareRepoReadSaveRequest.java:59)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitReadSaveService.getContent(GitReadSaveService.java:165)
	at io.jenkins.blueocean.rest.impl.pipeline.ScmResourceImpl.getContent(ScmResourceImpl.java:41)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
	... 94 more
```

cliffmeyers commented on blueocean-dashboard/src/main/js/credentials/git/GitCredentialsPicker.jsx:
```diff
@@ -0,0 +1,138 @@
+class GitCredentialsPicker extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {};
```
I think at least until we have a solid type system in place we should set initial values for all `state` keys, otherwise it's very hard to understand what the contract is.

kzantow replied:
👍

cliffmeyers commented on blueocean-dashboard/src/main/js/credentials/git/GitCredentialsPicker.jsx:
```diff
@@ -0,0 +1,138 @@
+                    <a target="jenkins-docs" href="https://jenkins.io/doc/book/blueocean/creating-pipelines/#creating-a-pipeline-for-a-git-repository">learn more</a>.
+                </p>
+                <FormElement>
+                    <textarea className="TextArea-control" ref="publicKey"
```
This style of ref usage is deprecated:  https://facebook.github.io/react/docs/refs-and-the-dom.html#legacy-api-string-refs

cliffmeyers replied:
Also could you expand on what made you use the raw `textarea` element versus the JDL `TextArea` control? If there is a shortcoming on the JDL control, we should file a ticket to improve it so we don't have to dup styles in `git.less`? I'm guessing it might be the copy/paste behavior?

kzantow replied:
There aren't complete duplicated styles in git.less but yes, it would be nice. For one it needed `readonly` but not `disabled` for clipboard copy to work right and the field to remain uneditable. It also needed a direct ref to the `textarea` component (although that could be found by traversing the DOM)

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCacheCloneReadSaveRequest.java:
```diff
@@ -0,0 +1,247 @@
+                                    if (user == null) {
+                                        throw new ServiceException.UnauthorizedException("Not authenticated");
+                                    }
+                                    credential = UserSSHKeyManager.getOrCreate(user);
```
@kzantow `UserSSHKeyManager.getOrCreate(user)` doesn't seem right here in case where credential id has been deleted or has not been setup then whats the point of creating a new ssh key pair/credential when the repo might not be setup to allow this new public key?

vivek replied:
Maybe doing `UserSSHKeyManager.get(user)` and then throw 428 if credential is null?

kzantow replied:
Well, presumably the user has viewed the UI, in which case the key would have been created and displayed, I don't see much reason to modify the behavior here.

vivek replied:
Umm, this is API and can be curled. API should return consistent error code, in this case that's expected behavior vs failing with some 500 error code.

kzantow replied:
This will attempt to connect with an SSH key if using an SSH URL. Whether the user has done something with it or not is irrelevant. If connecting fails, it'll return 428, not some 500 error.

vivek replied:
Great, but where does it throwing 428. I checked GitClient.clone() or its other method, they seem to throw RuntimeException. Other places in the code I see these are caught and 500 is returned.

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCacheCloneReadSaveRequest.java:
```diff
@@ -0,0 +1,247 @@
+                                commit.call();
+
+                                StandardCredentials credential = null;
+                                if (GitUtils.isSshUrl(gitSource.getRemote())) {
```
I see similar block of code other places, better to pull this in a util. Also, what expected if its not ssh url? Looks like follow up code will use null credential in that case which probably is ok security wise. For example if git server is setup with username/password, git operations might fail.

cliffmeyers commented on blueocean-dashboard/src/main/js/credentials/git/GitCredentialsPicker.jsx:
```diff
@@ -0,0 +1,138 @@
+    }
+}
+
+@observer
```
Is this component actually touching any `@observable` data structures?

kzantow replied:
heh.

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCloneReadSaveRequest.java:
```diff
@@ -0,0 +1,164 @@
+    @Override
+    byte[] read() throws IOException {
+        try {
+            cloneRepo();
```
`cloneRepo()` creates and sets `git` and `repositoryPath` non-final class instance variable. Its cleaner for `cloneRepo()` to return GitClient instance and remove these class instance variable. 

```java
GitClient git = cloneRepo();
...
```

In fact GitClient creation along with ssh credential search should be moved to a utility as I see similar code gets repeated at different places.

kzantow commented:
@cliffmeyers @michaelneale I botched the "normal" git ssh URL matching (e.g. it only worked for `ssh://...`), that's what you both ran into; it's fixed now - please give another try

kzantow commented:
@michaelneale to answer your q's - 1: @i386 wants to go only SSH key for git creation, and that's fine but I'd like to defer that until after we get beta-1 out; 3: once your key becomes invalid, it will prompt to add it

cliffmeyers commented:
@kzantow I'm getting the same error stack from the same repro step :/  This is a private Bitbucket repo with the following URL:  git@bitbucket.org:cliffmeyers/jenkins-test-private-empty.git

Slightly different error in the dialog:
![git-editor-1-error-loading-pipeline](https://user-images.githubusercontent.com/143466/30084193-5d704724-925f-11e7-81a5-0f1bcea2b0ed.png)

Server error:
```
java.lang.NullPointerException
	at jenkins.plugins.git.GitSCMFileSystem$1.invoke(GitSCMFileSystem.java:117)
	at jenkins.plugins.git.GitSCMFileSystem$1.invoke(GitSCMFileSystem.java:114)
	at jenkins.plugins.git.GitSCMFileSystem$3.invoke(GitSCMFileSystem.java:193)
	at org.jenkinsci.plugins.gitclient.AbstractGitAPIImpl.withRepository(AbstractGitAPIImpl.java:29)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.withRepository(CliGitAPIImpl.java:71)
	at jenkins.plugins.git.GitSCMFileSystem.invoke(GitSCMFileSystem.java:189)
	at jenkins.plugins.git.GitSCMFileSystem.<init>(GitSCMFileSystem.java:114)
	at jenkins.plugins.git.GitSCMFileSystem$BuilderImpl.build(GitSCMFileSystem.java:396)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:301)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:276)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitCacheCloneReadSaveRequest.getFilesystem(GitCacheCloneReadSaveRequest.java:145)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitBareRepoReadSaveRequest.read(GitBareRepoReadSaveRequest.java:58)
	at io.jenkins.blueocean.blueocean_git_pipeline.GitReadSaveService.getContent(GitReadSaveService.java:165)
	at io.jenkins.blueocean.rest.impl.pipeline.ScmResourceImpl.getContent(ScmResourceImpl.java:41)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
	at org.kohsuke.stapler.ForwardingFunction.invoke(ForwardingFunction.java:63)
	at io.jenkins.blueocean.commons.stapler.TreeResponse$Processor.invoke(TreeResponse.java:43)
	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:184)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:117)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:129)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:715)
Caused: javax.servlet.ServletException
```

cliffmeyers commented:
Reviewed my issue with @kzantow - that error was occurring when trying to edit against a totally empty repo with no commits. He is going to try to trap that error close to the source and report it back to the UI in some fashion.

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCloneReadSaveRequest.java:
```diff
@@ -56,7 +56,6 @@
     private final GitTool gitTool;
 
-    private GitClient git;
     private File repositoryPath;
```
This is not really needed as class instance variable.

vivek commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCloneReadSaveRequest.java:
```diff
@@ -0,0 +1,154 @@
+class GitCloneReadSaveRequest extends GitReadSaveRequest {
+    private final GitTool gitTool;
+
+    private File repositoryPath;
```
non final variable, not used, should be removed.

vivek commented:
LGTM 🐝

michaelneale commented:
looks good - one :ant: which is probably not related to this - but when I added https://github.com/multibranchorg/app-store-demo as a new git project - it discovered branches but never left this screen: 

![screen shot 2017-09-06 at 10 50 21 am](https://user-images.githubusercontent.com/14976/30089882-f7b77dfa-92f2-11e7-8c55-533eeb386cea.png)

I had to do a refresh. I expect this is not related though - probably just a "big" repo and indexing takes time... btu FYI. I would not let this get in the way of a :bee: as I tested other things. Nice one.

jmMeessen commented:
Thank you for pushing this and the work to get it merged.

scherler commented on blueocean-core-js/src/js/config.js:
```diff
@@ -7,6 +7,12 @@ const config = blueocean.config || {};
+// any all features added by ?features=SOMETHING,SOMETHING_ELSE
+const pfx = 'features=';
+const pfxlen = pfx.length;
+(window.location.href.split('?')[1] || '').split('&')
```
@kzantow that introduced a regression bug for unit testing...you are not catching a possible NPE!

`window.location` in unit testing is normally `undefined` and can be until `href` 

:bug:
